align/cell_fabric/canvas.py

```diff
@@ -143,6 +143,140 @@ def join_wires(self, wire, exclude_nets=None, include_nets=None, max_length=None
+
+        [mb, ma] = self.pdk[via.layer]['Stack']
+        assert mb is not None, f'Lower layer is not a metal'
+        assert ma is not None, f'Upper layer is not a metal'
+
+        # mh: horizontal wire, mv: vertical wire
+        if self.pdk[mb]['Direction'].upper() == 'H':
+            mh = self._find_generator(mb)
+            mv = self._find_generator(ma)
+        else:
+            mh = self._find_generator(ma)
+            mv = self._find_generator(mb)
+        mh_lines = self.rd.store_scan_lines[mh.layer]
+        mv_lines = self.rd.store_scan_lines[mv.layer]
+
+        via_matrix = self._construct_via_matrix(via)
+
+        for (mh_cl, mh_sl) in mh_lines.items():
+            for (_, mh_slr) in enumerate(mh_sl.rects):
+                mh_name = mh_slr.netName
+                if mh_name is None or mh_name in exclude_nets:
+                    continue
+                if include_nets is not None and mh_name not in include_nets:
+                    continue
+                for (mv_cl, mv_sl) in mv_lines.items():
+                    # Check only the scan lines that can intersect with ml_slr
+                    if mv_cl < 2*mh_slr.rect[0]:
+                        continue
+                    if mv_cl > 2*mh_slr.rect[2]:
+                        break
+                    for (_, mv_slr) in enumerate(mv_sl.rects):
+                        mv_name = mv_slr.netName
+                        if mv_name is None or mv_name != mh_name:
+                            continue
+                        # Check only the rectangles that overlap with the centerline
+                        if mh_cl > 2*mv_slr.rect[3]:
+                            continue
+                        if mh_cl < 2*mv_slr.rect[1]:
+                            break
+                        # Check if via exists
+                        if mh_cl in via_matrix and mv_cl in via_matrix[mh_cl]:
+                            continue
+                        # Check if DR-clean via can dropped
+                        self._drop_via_if_dr_clean(via, via_matrix, mh, mh_cl, mh_slr, mv, mv_cl, mv_slr)
+
+    def _find_generator(self, layer):
+        for key, gen in self.generators.items():
+            if gen.layer == layer:
+                return gen
+        assert False, f'A generator not found for {layer}'
+
+    def _construct_via_matrix(self, via):
+        via_lines = self.rd.store_scan_lines[via.layer]
+        via_matrix = dict()
+        for (_, via_sl) in via_lines.items():
+            for (_, via_slr) in enumerate(via_sl.rects):
+                mh_cl = via_slr.rect[0] + via_slr.rect[2]
+                mv_cl = via_slr.rect[1] + via_slr.rect[3]
+                if mh_cl not in via_matrix:
+                    via_matrix[mh_cl] = dict()
+                via_matrix[mh_cl][mv_cl] = via_slr.rect.copy()
+        return via_matrix
+
+    def _drop_via_if_dr_clean(self, via, via_matrix, mh, mh_cl, mh_slr, mv, mv_cl, mv_slr):
+        via_def = self.pdk[via.layer]
```

Comment:
You might want to assert if these divide by two are done on an odd number. (You are assuming they are even here, correct?)

Reply:
This assertion already exists in pdk.py:
`assert all(kwargs[x] % 2 == 0 for x in params[4:7]), f"One or more of {params[4:7]} in {kwargs} not a multiple of two"`